slick/src/main/scala/slick/jdbc/DatabaseUrlDataSource.scala

```diff
@@ -0,0 +1,88 @@
+package slick.jdbc
+
+import scala.language.reflectiveCalls
+
+import java.io.{PrintWriter, Closeable}
+import java.sql._
+import java.util.Properties
+import java.util.logging.Logger
+import javax.sql.DataSource
+
+import slick.SlickException
+import slick.util.{ClassLoaderUtil, Logging, ignoreFollowOnError}
+
+import scala.beans.BeanProperty
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+object DatabaseUrlDataSource {
+
+  private val PostgresFullUrl = "^postgres://([a-zA-Z0-9_]+):([^@]+)@([^/]+)/([^\\s]+)$".r
+  private val MysqlFullUrl = "^mysql://([a-zA-Z0-9_]+):([^@]+)@([^/]+)/([^\\s]+)$".r
+  private val MysqlCustomProperties = ".*\\?(.*)".r
+
+  def extractUrl(databaseUrl: Option[String]): String = {
+    extractUrlComponents(Some(databaseUrl.getOrElse(defaultUrl)))._1.orNull
+  }
+
+  def extractUser(databaseUrl: Option[String], default:String): String = {
+    extractUrlComponents(Some(databaseUrl.getOrElse(defaultUrl)))._2.map(_._1).getOrElse(default)
+  }
+
+  def extractPassword(databaseUrl: Option[String], default:String): String = {
+    extractUrlComponents(Some(databaseUrl.getOrElse(defaultUrl)))._2.map(_._2).getOrElse(default)
+  }
+
+  def extractUrlComponents(databaseUrl: Option[String]): (Option[String], Option[(String, String)]) = {
+    databaseUrl match {
+      case Some(PostgresFullUrl(username, password, host, dbname)) =>
+        Some(s"jdbc:postgresql://$host/$dbname") -> Some(username -> password)
+
+      case Some(url @ MysqlFullUrl(username, password, host, dbname)) =>
+        val defaultProperties = "?useUnicode=yes&characterEncoding=UTF-8&connectionCollation=utf8_general_ci"
+        val addDefaultPropertiesIfNeeded = MysqlCustomProperties.findFirstMatchIn(url).map(_ => "").getOrElse(defaultProperties)
+        Some(s"jdbc:mysql://$host/${dbname + addDefaultPropertiesIfNeeded}") -> Some(username -> password)
+
+      case Some(url) =>
+        Some(url) -> None
+
+      case None =>
+        None -> None
+    }
+  }
+
+  def defaultUrl():String = {
+    System.getenv("DATABASE_URL")
+  }
+}
+
+/** A DataSource that wraps the DriverManager API. It can be configured as a Java Bean and used
+  * both stand-alone and as a source for a connection pool. */
+class DatabaseUrlDataSource(
```

Comment:
Needs to have a no-args constructor to be used as a Java Bean. In fact, *only* a no-args contructor should be enough. DriverDataSource has the constructor with all fields to make it easier to use from Scala code but that's probably not a big concern here. I would have delegated to DriverDataSource instead of extending it. In retrospect, extending could be easier but I'm not sure if the current init() implementation is flexible enough. That's where you'd have to inject any new logic. Doing it in the constructor won't work.